htlcswitch/failure.go

```diff
@@ -43,6 +44,34 @@ func (f *ForwardingError) Error() string {
 	)
 }
 
+// Decode reads the bytes stream and converts it to the object. If the failure
+// message that the forwarding error wraps is serializable, its decode function
+// is used. Otherwise, the function is a nop.
+//
+// Note is is part of the lnwire.Serializable interface.
+func (f *ForwardingError) Decode(r io.Reader, pver uint32) error {
```

Comment:
Can't the caller just call `Decode` on `FailureMessage`? This suggests that all information is serialized, while in reality the detail gets lost.

Reply:
This is where my concern with `ForwardingError` embedding `lnwire.FailureMessage` comes in. They are mostly being passed around as `lnwire.FailuresMessage` and then we check whether it is a `ForwardingError`.

This gets buggy with `Serialize` because you have to check whether your`lnwire.FailureMessage` is a forwarding error, and unwrap it if you want to use the serialize code (since forwarding error doesn't implement Serialize). It's fine in one place but I worry that we will forget to unwrap somewhere else and then end up with bugs. 

Starting to think that an internal error type interface with `GetWireMessage() lnwire.FailureMessage` may be the way to go. To prevent having to duplicate every wire message we just have one type that is a wrapped wire message. 
```
type WrappedWire{
  msg lnwire.FailureMessage
}

func (w * WrappedWire) GetWireMessage() { return w.msg }
```

And then messages that are specific can return their associated wire failure:
```
type InsufficientBalanceError struct{}
func (i * InsufficientBalanceError) GetWireMessage() {return TemporaryChanFailure }
```

I like that this puts a barrier between internal errors and the wire by having to specifically call `GetWireMessage` and provides more info about failures. Thoughts?